flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala

```diff
@@ -70,24 +69,82 @@ class SinkTest extends TableTestBase {
   }
 
   @Test
-  def testCatalogTableSink(): Unit = {
-    val schemaBuilder = new TableSchema.Builder()
-    schemaBuilder.fields(Array("i"), Array(DataTypes.INT()))
-    val schema = schemaBuilder.build()
-    val sink = util.createCollectTableSink(schema.getFieldNames, Array(INT))
-    val catalog = Mockito.spy(new GenericInMemoryCatalog("dummy"))
-    val factory = Mockito.mock(classOf[TableSinkFactory[_]])
-    Mockito.when[Optional[_]](catalog.getTableFactory).thenReturn(Optional.of(factory))
-    Mockito.when[TableSink[_]](factory.createTableSink(
-      ArgumentMatchers.any(), ArgumentMatchers.any())).thenReturn(sink)
-    util.tableEnv.registerCatalog(catalog.getName, catalog)
-    util.tableEnv.useCatalog(catalog.getName)
-    val catalogTable = new CatalogTableImpl(schema, Map[String, String]().asJava, "")
-    catalog.createTable(new ObjectPath("default", "tbl"), catalogTable, false)
-    util.tableEnv.sqlQuery("select 1").insertInto("tbl")
+  def testTableSourceSinkFactory(): Unit = {
+    val factory = new TestContextTableFactory
+    util.tableEnv.getConfig.getConfiguration.setBoolean(factory.needContain, true)
+    util.tableEnv.registerCatalog("cat", new GenericInMemoryCatalog("default") {
```

Comment:
This only guarantees one of the paths, i.e. via `TableFactoryUtil#createTableSinkForCatalogTable`, which is only used by Hive. However, the another path is not covered, i.e. via `CatalogSourceTable#findAndCreateTableSource` which is used by most users.

Maybe you can upgrade the existing `TestCollectionTableFactory` to use the new interface and expose a session configuration, e.g. `collection.is-bounded`, the flag will be accessed via `Context#getConfiguration` and pass to the created `CollectionTableSource`.

Reply:
Yes, you are right, but I want to modify this step by step.
The previous method `createTableSink(ObjectPath tablePath, CatalogTable table)` is only work in `createTableSinkForCatalogTable` too.
after these commits looks good to you, I will create following commit to fix this.
What do you think?

Reply:
But without a test to cover the new interface, we have no idea whether it works as expected.

Reply:
What do you mean? Is this not a test cover?

Reply:
As I mentioned above, this test doesn't cover another path, i.e. `CatalogSourceTable#findAndCreateTableSource`.

Reply:
I'll modify these things by multi commits.